Fix ARM compilation breakage, introduce DeviceType enum, and document.

tim-seoss opened the pull request:
DRAFT PR for improving usability and documentation for the `Device::from_devnum` associated functions.

See #42 for background.

tim-seoss committed:
Fix compilation on platforms where `c_char` is not `i8`.

tim-seoss committed:
Introduce a `DeviceType` enum.

tim-seoss committed:
Use the DeviceType enum for the the `from_devnum` associated functions.

Since the underlying from_devnum functions have a somewhat confusing
API, provide extensive documentation for the correct usage for these
associated functions.

tim-seoss committed:
Expand documentation for the devtype associated function.

tim-seoss commented:
The use of num_enum (to derive `TryFrom`) might be kitchen-sinking it a bit, having searched around I can't find any (external or internal) APIs which return the b'c' and b'b' enum-substitute which the C libudev uses as a parameter.

Because of this, perhaps it should be removed, or perhaps feature-gated?

Drakulix commented on src/device.rs:
```diff
@@ -101,26 +118,95 @@ impl Device {
 
-    /// Creates a device for a given major/minor number
+    /// Creates a rust udev `Device` for a given UNIX device "special file" type and number. Uses
+    /// an existing [`Udev`] instance rather than creating one automatically.
```
I think this might be a copy-paste error? The `_with_context` variants use an existing instance.

Drakulix commented on src/device.rs:
````diff
@@ -21,6 +22,22 @@ pub struct Device {
+/// let d :DeviceType = b'c'.try_from()?;
+/// ```
+/// See also [`from_devnum`][crate::Device::from_devnum].
+#[derive(TryFromPrimitive)]
````
I think the additional dependency is not really worth this single derive. The enum is also super small, I don't think it would be huge burden for downstream to write a two-line `match`-statement, if anyone needs this.

Drakulix reviewed (commented):
Looks quite good to me!

Though with how detailed the docs are, I want to overhaul all of them now. :laughing:

tim-seoss commented:
Thanks for the review.  I've fixed up those commits and rebased.

[edit: Forgot to fixup formatting problems.  Done now.]

[edit: ...and the testing problem. Also done now.]

Drakulix reviewed (approved):
Thanks for working on this!
